plugins/inputs/intel_powerstat/services/msr.go

```diff
@@ -0,0 +1,297 @@
+	m.cpuCoresData[core].Mperf = newMperf
+	m.cpuCoresData[core].Aperf = newAperf
+	m.cpuCoresData[core].Tsc = newTsc
+	// MSR (1A2h) IA32_TEMPERATURE_TARGET bits 23:16.
+	m.cpuCoresData[core].ThrottleTemp = (newThrottleTemp >> 16) & 0xFF
+	// MSR (19Ch) IA32_THERM_STATUS bits 22:16.
+	m.cpuCoresData[core].Temp = (newTemp >> 16) & 0x7F
+
+	return nil
+}
+
+func (m *MsrServiceImpl) readValueFromFileAtOffset(ctx context.Context, ch chan uint64, reader io.ReaderAt, offset int64) error {
+	value, err := m.fs.ReadFileAtOffsetToUint64(reader, offset)
+	if err != nil {
+		return err
+	}
+
+	// Detect context cancellation and return an error if other goroutine fails
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case ch <- value:
+	}
+
+	return nil
+}
+
+// setCPUCores initialize cpuCoresData map.
+func (m *MsrServiceImpl) setCPUCores() error {
+	cpuPrefix := "cpu"
+	cpuCore := fmt.Sprintf("%s%s", cpuPrefix, "[0-9]*")
+	m.cpuCoresData = make(map[string]*data.MsrData)
+	cpuPaths, err := m.fs.GetStringsMatchingPatternOnPath(fmt.Sprintf("%s/%s", systemCPUPath, cpuCore))
+	if err != nil {
+		return err
+	}
+
+	for _, cpuPath := range cpuPaths {
+		core := strings.TrimPrefix(filepath.Base(cpuPath), cpuPrefix)
+		m.cpuCoresData[core] = &data.MsrData{
+			Mperf:        0,
+			Aperf:        0,
+			Tsc:          0,
+			C3:           0,
+			C6:           0,
+			C7:           0,
+			ThrottleTemp: 0,
+			Temp:         0,
+			MperfDelta:   0,
+			AperfDelta:   0,
+			TscDelta:     0,
+			C3Delta:      0,
+			C6Delta:      0,
+			C7Delta:      0,
+		}
+	}
+
+	return nil
+}
+
+func (m *MsrServiceImpl) getMsrOffsets() map[string]int64 {
+	return m.msrOffsets
+}
+
+func (m *MsrServiceImpl) calculateMsrOffsets() {
+	m.msrOffsets = map[string]int64{
+		c3StateResidencyLocation:           m.parseStringHexToInt(c3StateResidencyLocation),
+		c6StateResidencyLocation:           m.parseStringHexToInt(c6StateResidencyLocation),
+		c7StateResidencyLocation:           m.parseStringHexToInt(c7StateResidencyLocation),
+		maximumFrequencyClockCountLocation: m.parseStringHexToInt(maximumFrequencyClockCountLocation),
+		actualFrequencyClockCountLocation:  m.parseStringHexToInt(actualFrequencyClockCountLocation),
+		TimestampCounterLocation:           m.parseStringHexToInt(TimestampCounterLocation),
+		throttleTemperatureLocation:        m.parseStringHexToInt(throttleTemperatureLocation),
+		temperatureLocation:                m.parseStringHexToInt(temperatureLocation),
+	}
+}
+
+func (m *MsrServiceImpl) parseStringHexToInt(s string) int64 {
+	parsedInt, err := strconv.ParseInt(s, 0, 64)
+	if err != nil {
+		m.log.Errorf("error on parsing offset %s, err: %v", s, err)
+		return 0
+	}
+
+	return parsedInt
+}
+
+// NewMsrServiceWithFs returns new RaplServiceImpl struct with given FileService.
+func NewMsrServiceWithFs(logger telegraf.Logger, fs FileService) *MsrServiceImpl {
+	msrService := &MsrServiceImpl{
+		fs:  fs,
+		log: logger,
+	}
+	err := msrService.setCPUCores()
```

Comment:
The only location where this call can fail is if the well-formatted patterns passed to `filepath.Glob()` are broken. We then have a much deeper problem. So I'd rather return an error here.

Reply:
I understand, this is almost impossible to happen that this pattern is not well-formatted but for security reasons and for the sake of handling every error we would like to leave it as it is.